linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -1722,32 +1722,20 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
           val impls =
             if (useStaticResolution) List(staticCall(className, namespace, methodName))
             else dynamicCall(className, methodName)
-          val allocationSites =
-            (treceiver :: targs).map(_.tpe.allocationSite)
-          if (impls.isEmpty || impls.exists(impl =>
-              scope.implsBeingInlined((allocationSites, impl)))) {
-            // isEmpty could happen, have to leave it as is for the TypeError
-            treeNotInlined
-          } else if (impls.size == 1) {
-            val target = impls.head
-            val intrinsicCode = intrinsics(flags, target)
-            if (intrinsicCode >= 0) {
-              callIntrinsic(intrinsicCode, flags, Some(treceiver), methodName,
-                  targs, isStat, usePreTransform)(cont)
-            } else if (target.attributes.inlineable && (
-                target.attributes.shouldInline ||
-                shouldInlineBecauseOfArgs(target, treceiver :: targs))) {
-              /* When inlining a single method, the declared type of the `this`
-               * value is its enclosing class.
-               */
-              val receiverType = receiverTypeFor(target)
-              inline(allocationSites, Some((receiverType, treceiver)), targs,
-                  target, isStat, usePreTransform)(cont)
-            } else {
+          if (impls.size == 1) {
+            pretransformSingleDispatch(flags, impls.head, Some(treceiver), targs, isStat, usePreTransform)(cont) {
               treeNotInlined
```

Comment:
This makes me wonder: Would it make sense in this case to emit `ApplyStatically` instead of `Apply`?

This would allow us to simplify the call flow: Instead of providing `treeNotInlined` we could build it inside `pretransformSingleDispatch`.

Just TBC: I don't think this should go into this PR.

Reply:
Not really. We tried this and much more in a PR to use `ApplyStatically` everywhere possible, but the results were inconclusive in terms of performance. Some benchmarks were faster, some slower. The takeaway was that it's better to preserve `Apply` nodes as is, overall.